data/modules/FlightLog/FlightLog.lua

```diff
@@ -0,0 +1,326 @@
+local skip_first_docking = nil
+
+-- default values (private)
+---@type integer
+local MaxTotalDefaultElements = 3000
+
+-- private data - the log itself
+---@type FlightLogEntry.Base[]
+local FlightLogData = {}
+
+local FlightLog
+FlightLog = {
+
+--
+-- Group: Methods
+--
+--
+--
+-- Method: MakeCustomEntry
+--
+-- Create a custom entry. A set of information is automatically
+-- compiled, in a header.
+--
+-- > FlightLog.MakeCustomEntry(text)
+--
+-- Header:
+--
+--   path - System path, pointing to player's current sytem
+--   time - Game date
+--   money - Financial balance at time of record creation
+--   location - Array, with two strings: flight state, and relevant additional string
+--   manual entry - Free text string
+--
+-- Parameters:
+--
+--   text - Text to accompany the log
+--
+
+	MakeCustomEntry = function (text)
+		text = text or ""
+		local location = ""
+		local state = Game.player:GetFlightState()
+		local path = ""
+
+		if state == "DOCKED" then
+			local station = Game.player:GetDockedWith()
+			local parent_body = station.path:GetSystemBody().parent.name
+			location = {station.type, station.label, parent_body}
+			path = Game.system.path
+		elseif state == "DOCKING" or state == "UNDOCKING" then
+			location = {state, Game.player:FindNearestTo("SPACESTATION").label}
+			path = Game.system.path
+		elseif state == "FLYING" then
+			if Game.player.frameBody then
+				location = {state, Game.player.frameBody.label}
+			else
+				location = {state, Game.system.name} -- if orbiting a system barycenter, there will be no frame object
+			end
+			path = Game.system.path
+		elseif state == "LANDED" then
+			path = Game.system.path
+			local alt, vspd, lat, long = Game.player:GetGPS()
+			if not (lat and long) then
+				lat, long = "nil", "nil"
+			end
+			location = {state, Game.player:FindNearestTo("PLANET").label, lat, long}
+		elseif state == "JUMPING" or state == "HYPERSPACE" then
+			--if in hyperspace, there's no Game.system
+			local spath, sysname = Game.player:GetHyperspaceDestination()
+			path = spath
+			location = {state, sysname}
+		end
+
+		table.insert(FlightLogData,1, FlightLogEntry.Custom.New( path, Game.time, Game.player:GetMoney(), location, text ) )
+	end,
+
+}
+
+
+
+function FlightLog.SkipFirstDocking()
+	skip_first_docking = true
+end
+
+--- Method: GetLogEntries
+---@param types table[string,boolean]|nil  Keys are log types to include, set the boolean to true for the ones you want
+---@param maximum integer|nil              Maximum number of entries to include
+---@param earliest_first boolean|nil       Should the log start with the oldest entry or the most recent
+--- 
+---@return function():FlightLogEntry.Base  An iterator function that when called repeatedly returns the next entry or nil when complete
+---
+--- Example:
+---
+--- > for entry in FlightLog.GetLogEntries( { "Custom", "System", "Station" ) do
+--- >   print( entry.GetType(), entry.entry )
+--- > end
```

Comment:
OK, that's good to know, this was my first foray into any kind of parameter hinting in Lua and I pretty much learned as I went along so I understand it's probably a bit rough.

@Web-eWorks , if I understand correctly, to do in entirety what you're suggesting here I'd need to have the following:

```
-- Method: GetLogEntries
--
-- Example:
--
-- > for entry in FlightLog.GetLogEntries( { "Custom", "System", "Station" ) do
-- >   print( entry.GetType(), entry.entry )
-- > end
--
-- Parameters:
-- 
--    types         A table where keys are log types to include and the value is a boolean, set to true to include that type
--    maximum   An optional integer, with the maximum number of entries to include
--    earliest_first An optional boolean to say if the log should be ordered in chronological order or reverse chronological order
-- 
-- Returns:
--
--    An iterator function that when called repeatedly returns the next entry or nil when complete
--
--
---@param types table[string,boolean]?     Keys are log types to include, set the boolean to true for the ones you want
---@param maximum integer?                 Maximum number of entries to include
---@param earliest_first boolean?          Should the log start with the oldest entry or the most recent
--- 
---@return fun():FlightLogEntry.Base       An iterator function that when called repeatedly returns the next entry or nil when complete
```

Reply:
I personally don't add descriptions to the `---@param` annotations for public functions which are also documented via NaturalDocs, as that's duplicated effort.

Otherwise that looks fairly correct - although the Natural Docs parameters list needs to be formatted with a `-` to separate the parameter name from the description. We sometimes prefer to include the type of the parameter in the start of the description string to provide extra information to people reading the compiled codedoc:

```
types - [type description], human description
```
e.g.
```
types - optional table<string, boolean>, controls which flight log types should be included in the iterator results
```

Reply:
Additionally, `table<string, boolean>` is a key-value table type annotation that will be parsed correctly by the Lua language server. `table[string, boolean]` is a invalid type annotation and the type will be deduced as a general-purpose `table` value followed by a comment string. Square brackets are only used in array-of-type annotations as e.g. `string[]`.

Reply:
OK, this should also be sorted out now